From: Suren Baghdasaryan <surenb@google.com>
Subject: alloc_tag: use release_pages() in the cleanup path
Date: Tue, 9 Sep 2025 16:34:07 -0700

Patch series "Minor fixes for memory allocation profiling".

Over the last couple months I gathered a few reports of minor issues in
memory allocation profiling which are addressed in this patchset.


This patch (of 3):

When bulk-freeing an array of pages use release_pages() instead of freeing
them page-by-page:

Link: https://lkml.kernel.org/r/20250909233409.1013367-1-surenb@google.com
Link: https://lkml.kernel.org/r/20250909233409.1013367-2-surenb@google.com
Suggested-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Suren Baghdasaryan <surenb@google.com>
Cc: Brendan Jackman <jackmanb@google.com>
Cc: David Wang <00107082@163.com>
Cc: Johannes Weiner <hannes@cmpxchg.org>
Cc: Kent Overstreet <kent.overstreet@linux.dev>
Cc: Michal Hocko <mhocko@suse.com>
Cc: Pasha Tatashin <pasha.tatashin@soleen.com>
Cc: Shakeel Butt <shakeel.butt@linux.dev>
Cc: Sourav Panda <souravpanda@google.com>
Cc: Usama Arif <usamaarif642@gmail.com>
Cc: Vlastimil Babka <vbabka@suse.cz>
Cc: Zi Yan <ziy@nvidia.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/alloc_tag.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/lib/alloc_tag.c~alloc_tag-use-release_pages-in-the-cleanup-path
+++ a/lib/alloc_tag.c
@@ -438,9 +438,10 @@ static int vm_module_tags_populate(void)
 		if (nr < more_pages ||
 		    vmap_pages_range(phys_end, phys_end + (nr << PAGE_SHIFT), PAGE_KERNEL,
 				     next_page, PAGE_SHIFT) < 0) {
+			release_pages_arg arg = { .pages = next_page };
+
 			/* Clean up and error out */
-			for (int i = 0; i < nr; i++)
-				__free_page(next_page[i]);
+			release_pages(arg, nr);
 			return -ENOMEM;
 		}
 
_
